test/python/dialects/indexing/dialect.py

```diff
@@ -439,3 +440,37 @@ def testSimpleLiteralIndexing():
     print(Tensor(w.owner.operands[1]))
     # CHECK: %{{.*}} = indexing.gather %[[TEN]][%[[CST0]]] gather_dims([0, 1, 2]) unique : (tensor<10x10x10x10xi32>, tensor<3xindex>) -> tensor<10xi32>
     print(w.owner)
+
+
+# CHECK-LABEL: TEST: testCanonicalizeTupleIndexCastListLiteral
+@run
+def testCanonicalizeTupleIndexCastListLiteral():
+  with mlir_mod_ctx() as module:
+
+    for n_tens in range(1, 4):
+      uniqs = set()
+      n_slices = 4 - n_tens
+      ten_idx = [[0], [1]]
+      slice_idx = slice(None)
+      for p in permutations([str(ten_idx)] * n_tens +
+                            [str(slice_idx)] * n_slices):
+        uniqs.add(p)
+
+      for u in uniqs:
+        u = tuple(u)
+        tens_is = [i for i, t in enumerate(u) if t == str(ten_idx)]
+        slice_is = [i for i, s in enumerate(u) if s == str(slice_idx)]
+
+        tens_slices = _canonicalize_tuple_index(tuple(map(eval, u)), 4)
```

Comment:
Can you try to find a way to do this without `eval`?

Reply:
You can't - `slice(None)` and `[[0], [1]]` objects aren't hashable (but their `str` representations are) so you wouldn't be able to find unique permutations.